File: bindings/python/examples/how_tos/alias/governance_transition.py
import os

from dotenv import load_dotenv

from iota_sdk import Wallet, FilterOptions, Utils, UnlockConditionType, StateControllerAddressUnlockCondition

load_dotenv()

# In this example we will update the state controller of an alias output.

for env_var in ['WALLET_DB_PATH', 'STRONGHOLD_PASSWORD', 'EXPLORER_URL']:
    if env_var not in os.environ:
        raise Exception(f".env {env_var} is undefined, see .env.example")

wallet = Wallet(os.environ['WALLET_DB_PATH'])

account = wallet.get_account('Alice')

# Sync account with the node
balance = account.sync()

if len(balance.aliases) == 0:
    raise Exception("No Alias available in account 'Alice'")

alias_id = balance.aliases[0]

alias_output_data = account.unspent_outputs(
    FilterOptions(aliasIds=[alias_id]))[0]
print(f"Alias {alias_id} found in unspent output: {alias_output_data.outputId}")

wallet.set_stronghold_password(os.environ["STRONGHOLD_PASSWORD"])

print(f"{ account.generate_ed25519_addresses(1)[0].address}")
new_state_controller = Utils.parse_bech32_address(
    account.generate_ed25519_addresses(1)[0].address)
print(f"{new_state_controller.__dict__}")

alias_output = alias_output_data.output


def update_state_controller(unlock_condition):
    """
    Replace the address in the StateControllerAddressUnlockCondition
    """
    if unlock_condition.type == UnlockConditionType.StateControllerAddress:
        return StateControllerAddressUnlockCondition(new_state_controller)
    return unlock_condition


updated_unlock_conditions = list(map(
    update_state_controller, alias_output.unlockConditions))
updated_alias_output = wallet.get_client().build_alias_output(
    alias_id,
    unlock_conditions=updated_unlock_conditions,
    state_index=alias_output.stateIndex,
    state_metadata=alias_output.stateMetadata,
    foundry_counter=alias_output.foundryCounter,
    immutable_features=alias_output.immutableFeatures,
    features=alias_output.features,
)

print('Sending transaction...')
transaction = account.send_outputs([updated_alias_output])
print(f'Transaction sent: {transaction.transactionId}')

# Wait for transaction to get included
blockId = account.retry_transaction_until_included(
    transaction.transactionId)
print(f'Block included: {os.environ["EXPLORER_URL"]}/block/{blockId}')

File: bindings/nodejs/examples/how_tos/alias/governance-transition.ts
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {
    AliasOutput,
    StateControllerAddressUnlockCondition,
    UnlockConditionType,
    Utils,
    Wallet,
    initLogger,
} from '@iota/sdk';

// This example uses secrets in environment variables for simplicity which should not be done in production.
require('dotenv').config({ path: '.env' });

// Run with command:
// yarn run-example ./how_tos/alias/governance-transition.ts

// In this example we will update the state controller of an alias output.
async function run() {
    initLogger();
    for (const envVar of [
        'WALLET_DB_PATH',
        'STRONGHOLD_PASSWORD',
        'EXPLORER_URL',
    ]) {
        if (!(envVar in process.env)) {
            throw new Error(`.env ${envVar} is undefined, see .env.example`);
        }
    }

    try {
        // Create the wallet
        const wallet = new Wallet({
            storagePath: process.env.WALLET_DB_PATH,
        });

        // Get the account we generated with `01-create-wallet`
        const account = await wallet.getAccount('Alice');

        // May want to ensure the account is synced before sending a transaction.
        const balance = await account.sync();

        if (balance.aliases.length == 0) {
            throw new Error(`No Alias available in account 'Alice'`);
        }

        // We try to update the first alias in the account
        const aliasId = balance.aliases[0];

        const aliasOutputData = (
            await account.unspentOutputs({ aliasIds: [aliasId] })
        )[0];
        console.log(
            `Alias ${aliasId} found in unspent output: ${aliasOutputData.outputId}`,
        );

        await wallet.setStrongholdPassword(
            process.env.STRONGHOLD_PASSWORD as string,
        );

        const newStateController = Utils.parseBech32Address(
            (await account.generateEd25519Addresses(1))[0].address,
        );

        const aliasOutput = aliasOutputData.output as AliasOutput;
        const updatedUnlockConditions = aliasOutput.unlockConditions.map(
            (unlock) => {
                if (unlock.type == UnlockConditionType.StateControllerAddress) {
                    return new StateControllerAddressUnlockCondition(
                        newStateController,
                    );
                } else {
                    return unlock;
                }
            },
        );

        const updatedAliasOutput = await (
            await wallet.getClient()
        ).buildAliasOutput({
            aliasId,
            amount: aliasOutput.amount,
            unlockConditions: updatedUnlockConditions,
            stateIndex: aliasOutput.stateIndex,
            stateMetadata: aliasOutput.stateMetadata,
            foundryCounter: aliasOutput.foundryCounter,
            immutableFeatures: aliasOutput.immutableFeatures,
            features: aliasOutput.features,
        });

        console.log('Sending transaction...');

        const transaction = await account.sendOutputs([updatedAliasOutput]);
        console.log(`Transaction sent: ${transaction.transactionId}`);

        // Wait for transaction to get included
        const blockId = await account.retryTransactionUntilIncluded(
            transaction.transactionId,
        );
        console.log(
            `Block included: ${process.env.EXPLORER_URL}/block/${blockId}`,
        );
    } catch (error) {
        console.log('Error: ', error);
    }
    process.exit(0);
}

run();
